pkg/serviceaccounttoken/secret.go

```diff
@@ -5,41 +5,61 @@ import (
 	"fmt"
 	"time"
 
+	corecontrollers "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
 	"github.com/sirupsen/logrus"
 	v1 "k8s.io/api/core/v1"
-	apierror "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes"
 )
 
-const serviceAccountSecretAnnotation = "kubernetes.io/service-account.name"
+const (
+	// ServiceAccountSecretLabel is the label used to search for the secret belonging to a service account.
+	ServiceAccountSecretLabel = "cattle.io/service-account.name"
 
-// secretGetter is an abstraction over any kind of secret getter.
+	serviceAccountSecretAnnotation = "kubernetes.io/service-account.name"
+)
+
+// secretLister is an abstraction over any kind of secret lister.
 // The caller can use any cache or client it has available, whether that is from norman, wrangler, or client-go,
 // as long as it can wrap it in a simplified lambda with this signature.
-type secretGetter func(namespace, name string) (*v1.Secret, error)
+type secretLister func(namespace string, selector labels.Selector) ([]*v1.Secret, error)
 
 // EnsureSecretForServiceAccount gets or creates a service account token Secret for the provided Service Account.
 // For k8s <1.24, the secret is automatically generated for the service account. For >=1.24, we need to generate it explicitly.
-func EnsureSecretForServiceAccount(ctx context.Context, secretGetter secretGetter, clientSet kubernetes.Interface, sa *v1.ServiceAccount) (*v1.Secret, error) {
-	if secretGetter == nil {
-		secretGetter = func(namespace, name string) (*v1.Secret, error) {
-			return clientSet.CoreV1().Secrets(namespace).Get(ctx, name, metav1.GetOptions{})
-		}
-	}
+func EnsureSecretForServiceAccount(ctx context.Context, secretsCache corecontrollers.SecretCache, clientSet kubernetes.Interface, sa *v1.ServiceAccount) (*v1.Secret, error) {
```

Comment:
The `corecontrollers` package that you are using here has a `corecontrollers.SecretController` which provides a way to both read/write these objects. If you can use that as input, it would make it not necessary to pass in the clientSet, which I think would be optimal.

Reply:
That would work for most places, but some things (https://github.com/rancher/rancher/blob/release/v2.8/pkg/agent/cluster/cluster.go) that use this don't have a wrangler context but still need to try to access the k8s API, that's why I have both here.